src/vm-vcpu/src/vcpu/mod.rs

```diff
@@ -244,81 +245,91 @@ impl KvmVcpu {
             self.running = true;
         }
 
-        match self.vcpu_fd.run() {
-            Ok(exit_reason) => {
-                match exit_reason {
-                    VcpuExit::Shutdown | VcpuExit::Hlt => {
-                        println!("Guest shutdown: {:?}. Bye!", exit_reason);
-                        if stdin().lock().set_canon_mode().is_err() {
-                            eprintln!("Failed to set canon mode. Stdin will not echo.");
+        loop {
+            match self.vcpu_fd.run() {
+                Ok(exit_reason) => {
+                    match exit_reason {
+                        VcpuExit::Shutdown | VcpuExit::Hlt => {
+                            println!("Guest shutdown: {:?}. Bye!", exit_reason);
+                            if stdin().lock().set_canon_mode().is_err() {
```

Comment:
I think this should be moved in the vmm code because
* the vcpu shouldn't care about the terminal's mode, it's an application setting
* it's not even needed if the serial console is absent, or has a different input sink than stdin (not the case for the present reference VMM but should still be configurable, I'll add an issue)

Previously there was no other chance to reset the terminal settings because the process exited abruptly here.

Reply:
I was also thinking that the exit handling should be part of the VMM. What I was planning on doing is a bit different though. I think even what vcpu exists we want to handle is the responsibility of the VMM, and was thinking about moving the full processing there. A try of doing so is in one of my branches: https://github.com/andreeaflorescu/vmm-reference/commit/000b41fe257da6b488c09aef6004ff972ecf2f66

I pushed this change a bit more into the future because the implementation of clean exit & pause resume will infer changes into the exit handler as well.

Would you mind leaving it as is for now, and re-discuss this when I open the other PR that treats all the vcpu exits outside of this crate?